File: forge-core/internal/kerneloperationalcontract/graph.go
package kerneloperationalcontract

import (
	"fmt"
	"reflect"
	"sort"
)

func invocationReference(value CapabilityInvocation) CapabilityInvocationRef {
	return CapabilityInvocationRef{InvocationID: value.InvocationID,
		InvocationSHA256: value.InvocationSHA256}
}

func artifactReceiptReference(value ArtifactReceipt) ArtifactReceiptRef {
	return ArtifactReceiptRef{ArtifactReceiptID: value.ArtifactReceiptID,
		ArtifactReceiptSHA256: value.ArtifactReceiptSHA256}
}

func eventReference(value InteractionEvent) InteractionEventRef {
	return InteractionEventRef{EventID: value.EventID, EventSHA256: value.EventSHA256}
}

func executionReference(value ExecutionReceipt) ExecutionReceiptRef {
	return ExecutionReceiptRef{ExecutionReceiptID: value.ExecutionReceiptID,
		ExecutionReceiptSHA256: value.ExecutionReceiptSHA256}
}

func validateCommonContext(value *KernelOperationalReferenceClosure) error {
	base := value.CapabilityInvocations[0]
	for index := range value.ArtifactReceipts {
		if !reflect.DeepEqual(value.ArtifactReceipts[index].Bindings, base.Bindings) ||
			!reflect.DeepEqual(value.ArtifactReceipts[index].TaskBinding, base.TaskBinding) {
			return fmt.Errorf("every record must carry identical bindings and TaskBinding")
		}
	}
	for index := range value.CapabilityInvocations {
		if !reflect.DeepEqual(value.CapabilityInvocations[index].Bindings, base.Bindings) ||
			!reflect.DeepEqual(value.CapabilityInvocations[index].TaskBinding, base.TaskBinding) {
			return fmt.Errorf("every record must carry identical bindings and TaskBinding")
		}
	}
	for index := range value.InteractionEvents {
		if !reflect.DeepEqual(value.InteractionEvents[index].Bindings, base.Bindings) ||
			!reflect.DeepEqual(value.InteractionEvents[index].TaskBinding, base.TaskBinding) {
			return fmt.Errorf("every record must carry identical bindings and TaskBinding")
		}
	}
	for index := range value.ExecutionReceipts {
		if !reflect.DeepEqual(value.ExecutionReceipts[index].Bindings, base.Bindings) ||
			!reflect.DeepEqual(value.ExecutionReceipts[index].TaskBinding, base.TaskBinding) {
			return fmt.Errorf("every record must carry identical bindings and TaskBinding")
		}
	}
	return nil
}

func sameRetryStatic(first, next CapabilityInvocation) bool {
	return reflect.DeepEqual(first.Bindings, next.Bindings) &&
		reflect.DeepEqual(first.Capability, next.Capability) &&
		reflect.DeepEqual(first.CapabilityGrantRef, next.CapabilityGrantRef) &&
		first.CorrelationID == next.CorrelationID &&
		reflect.DeepEqual(first.DeclaredOutputSlots, next.DeclaredOutputSlots) &&
		first.IdempotencyKey == next.IdempotencyKey &&
		reflect.DeepEqual(first.InputArtifactReceiptRefs, next.InputArtifactReceiptRefs) &&
		first.RequestedActionSHA256 == next.RequestedActionSHA256 &&
		reflect.DeepEqual(first.Subject, next.Subject) &&
		reflect.DeepEqual(first.TaskBinding, next.TaskBinding)
}

func validateRetryChain(value *KernelOperationalReferenceClosure) error {
	if len(value.CapabilityInvocations) != len(value.ExecutionReceipts) {
		return fmt.Errorf("each invocation requires exactly one ExecutionReceipt")
	}
	base := value.CapabilityInvocations[0]
	for index := range value.CapabilityInvocations {
		invocation := value.CapabilityInvocations[index]
		receipt := value.ExecutionReceipts[index]
		expectedAttempt := int64(index + 1)
		if invocation.Attempt != expectedAttempt || receipt.Attempt != expectedAttempt {
			return fmt.Errorf("invocations and receipts must be ordered contiguous attempts 1..N")
		}
		if !sameRetryStatic(base, invocation) {
			return fmt.Errorf("retry invocations must preserve the complete static request")
		}
		var prior *ExecutionReceiptRef
		if index > 0 {
			item := executionReference(value.ExecutionReceipts[index-1])
			prior = &item
		}
		if !reflect.DeepEqual(invocation.PriorExecutionReceiptRef, prior) ||
			!reflect.DeepEqual(receipt.PriorExecutionReceiptRef, prior) {
			return fmt.Errorf("prior receipt must be the preceding attempt")
		}
		if index > 0 && value.ExecutionReceipts[index-1].Outcome == "succeeded" {
			return fmt.Errorf("a succeeded execution cannot be retried")
		}
		if index > 0 && invocation.RequestedAtUnixMS < value.ExecutionReceipts[index-1].EndedAtUnixMS {
			return fmt.Errorf("retry request precedes the prior attempt end")
		}
	}
	return nil
}

func artifactReceiptMap(values []ArtifactReceipt) (map[string]ArtifactReceipt, error) {
	result := make(map[string]ArtifactReceipt, len(values))
	for _, value := range values {
		if _, exists := result[value.ArtifactReceiptID]; exists {
			return nil, fmt.Errorf("duplicate ArtifactReceipt identity")
		}
		result[value.ArtifactReceiptID] = value
	}
	return result, nil
}

func resolveInputs(invocation CapabilityInvocation, receipt ExecutionReceipt,
	available map[string]ArtifactReceipt, used map[string]bool) error {
	artifacts := make([]ArtifactRef, 0, len(invocation.InputArtifactReceiptRefs))
	seen := make(map[ArtifactRef]bool)
	for _, reference := range invocation.InputArtifactReceiptRefs {
		member, found := available[reference.ArtifactReceiptID]
		if !found || reference != artifactReceiptReference(member) {
			return fmt.Errorf("input ArtifactReceipt reference is unresolved")
		}
		if member.ReceiptRole != "declared_input" ||
			member.CreatedAtUnixMS > invocation.RequestedAtUnixMS {
			return fmt.Errorf("invocation input role or time is invalid")
		}
		if seen[member.Artifact] {
			return fmt.Errorf("invocation inputs must project distinct ArtifactRefs")
		}
		seen[member.Artifact], used[member.ArtifactReceiptID] = true, true
		artifacts = append(artifacts, member.Artifact)
	}
	sortArtifacts(artifacts)
	if !reflect.DeepEqual(receipt.InputArtifacts, artifacts) {
		return fmt.Errorf("ExecutionReceipt input_artifacts must exactly project inputs")
	}
	return nil
}

func outputsFor(invocation CapabilityInvocation, receipt ExecutionReceipt,
	values []ArtifactReceipt, used map[string]bool) error {
	reference := invocationReference(invocation)
	outputs := make([]ArtifactReceipt, 0)
	for _, value := range values {
		if value.ProducerInvocationRef != nil && *value.ProducerInvocationRef == reference {
			outputs = append(outputs, value)
		}
	}
	slots := make([]string, len(outputs))
	refs := make([]ArtifactReceiptRef, len(outputs))
	for index, output := range outputs {
		slots[index], refs[index] = output.Slot, artifactReceiptReference(output)
		used[output.ArtifactReceiptID] = true
		if output.CreatedAtUnixMS < receipt.StartedAtUnixMS ||
			output.CreatedAtUnixMS > receipt.EndedAtUnixMS {
			return fmt.Errorf("output ArtifactReceipt time is outside its execution")
		}
	}
	sort.Strings(slots)
	if hasDuplicate(slots) || !reflect.DeepEqual(slots, invocation.DeclaredOutputSlots) {
		return fmt.Errorf("declared_output_slots must exactly cover output receipts")
	}
	sort.Slice(refs, func(i, j int) bool { return refs[i].ArtifactReceiptID < refs[j].ArtifactReceiptID })
	if !reflect.DeepEqual(refs, receipt.OutputArtifactReceiptRefs) {
		return fmt.Errorf("ExecutionReceipt must reference every exact output receipt")
	}
	return nil
}

func eventsFor(invocation CapabilityInvocation, receipt ExecutionReceipt,
	values []InteractionEvent) ([]InteractionEvent, error) {
	reference := invocationReference(invocation)
	members := make([]InteractionEvent, 0)
	for _, value := range values {
		if value.InvocationRef == reference {
			members = append(members, value)
		}
	}
	refs := make([]InteractionEventRef, len(members))
	for index, event := range members {
		if event.LogicalSequence != int64(index+1) {
			return nil, fmt.Errorf("events must be contiguous sequences per invocation")
		}
		refs[index] = eventReference(event)
		var cause *InteractionEventRef
		if index > 0 {
			cause = &refs[index-1]
		}
		if !reflect.DeepEqual(event.CausationEventRef, cause) {
			return nil, fmt.Errorf("event causation must be immediately preceding")
		}
		if event.OccurredAtUnixMS < receipt.StartedAtUnixMS ||
			event.OccurredAtUnixMS > receipt.EndedAtUnixMS {
			return nil, fmt.Errorf("event time is outside its execution")
		}
		if index > 0 && event.OccurredAtUnixMS < members[index-1].OccurredAtUnixMS {
			return nil, fmt.Errorf("event times must be nondecreasing by logical sequence")
		}
	}
	if !reflect.DeepEqual(refs, receipt.EventRefs) {
		return nil, fmt.Errorf("ExecutionReceipt event_refs must exactly cover its events")
	}
	return members, nil
}

func validateAttempt(invocation CapabilityInvocation, receipt ExecutionReceipt,
	artifactMap map[string]ArtifactReceipt, artifacts []ArtifactReceipt,
	events []InteractionEvent, used map[string]bool) ([]InteractionEvent, error) {
	if receipt.InvocationRef != invocationReference(invocation) {
		return nil, fmt.Errorf("ExecutionReceipt must reference its matching invocation")
	}
	if receipt.CorrelationID != invocation.CorrelationID {
		return nil, fmt.Errorf("receipt and invocation correlation_id differ")
	}
	if receipt.StartedAtUnixMS < invocation.RequestedAtUnixMS {
		return nil, fmt.Errorf("execution starts before its invocation request")
	}
	if err := resolveInputs(invocation, receipt, artifactMap, used); err != nil {
		return nil, err
	}
	if err := outputsFor(invocation, receipt, artifacts, used); err != nil {
		return nil, err
	}
	members, err := eventsFor(invocation, receipt, events)
	if err != nil {
		return nil, err
	}
	for _, event := range members {
		if event.CorrelationID != invocation.CorrelationID {
			return nil, fmt.Errorf("event and invocation correlation_id differ")
		}
	}
	return members, nil
}

func validateArtifactInventory(value *KernelOperationalReferenceClosure) error {
	created := make(map[ArtifactRef][]int64)
	projected := make(map[ArtifactRef]bool)
	for _, receipt := range value.ArtifactReceipts {
		created[receipt.Artifact] = append(created[receipt.Artifact], receipt.CreatedAtUnixMS)
		projected[receipt.Artifact] = true
	}
	for _, event := range value.InteractionEvents {
		for _, artifact := range event.ArtifactRefs {
			if !hasNonfutureReceipt(created[artifact], event.OccurredAtUnixMS) {
				return fmt.Errorf("Event ArtifactRef needs a non-future included ArtifactReceipt")
			}
			projected[artifact] = true
		}
	}
	for _, receipt := range value.ExecutionReceipts {
		for _, artifact := range receipt.InputArtifacts {
			projected[artifact] = true
		}
	}
	expected := make([]ArtifactRef, 0, len(projected))
	for artifact := range projected {
		expected = append(expected, artifact)
	}
	sortArtifacts(expected)
	if !reflect.DeepEqual(expected, value.Artifacts) {
		return fmt.Errorf("closure artifacts must exactly equal all referenced ArtifactRefs")
	}
	return nil
}

func hasNonfutureReceipt(times []int64, occurred int64) bool {
	for _, created := range times {
		if created <= occurred {
			return true
		}
	}
	return false
}

func sortArtifacts(values []ArtifactRef) {
	sort.Slice(values, func(i, j int) bool {
		left, _ := canonicalKey(values[i])
		right, _ := canonicalKey(values[j])
		return string(left) < string(right)
	})
}

func validateReferenceGraph(value *KernelOperationalReferenceClosure) error {
	if err := validateCommonContext(value); err != nil {
		return err
	}
	if err := validateRetryChain(value); err != nil {
		return err
	}
	artifactMap, err := artifactReceiptMap(value.ArtifactReceipts)
	if err != nil {
		return err
	}
	used := make(map[string]bool)
	flattened := make([]InteractionEvent, 0, len(value.InteractionEvents))
	for index := range value.CapabilityInvocations {
		members, err := validateAttempt(value.CapabilityInvocations[index],
			value.ExecutionReceipts[index], artifactMap, value.ArtifactReceipts,
			value.InteractionEvents, used)
		if err != nil {
			return err
		}
		flattened = append(flattened, members...)
	}
	if !reflect.DeepEqual(flattened, value.InteractionEvents) {
		return fmt.Errorf("interaction_events must be ordered by attempt then sequence")
	}
	if len(used) != len(artifactMap) {
		return fmt.Errorf("ArtifactReceipt inventory contains an orphan")
	}
	return validateArtifactInventory(value)
}

File: forge-runtime/crates/domain/src/kernel_operational_contract/graph.rs
use std::collections::{BTreeSet, HashMap};

use super::{
    ArtifactReceipt, ArtifactReceiptRef, ArtifactRef, CapabilityInvocation,
    CapabilityInvocationRef, ExecutionReceipt, ExecutionReceiptRef, InteractionEvent,
    InteractionEventRef, KernelOperationalContractError, KernelOperationalReferenceClosure,
    invalid, wire,
};

fn invocation_ref(value: &CapabilityInvocation) -> CapabilityInvocationRef {
    CapabilityInvocationRef {
        invocation_id: value.invocation_id.clone(),
        invocation_sha256: value.invocation_sha256.clone(),
    }
}

fn artifact_receipt_ref(value: &ArtifactReceipt) -> ArtifactReceiptRef {
    ArtifactReceiptRef {
        artifact_receipt_id: value.artifact_receipt_id.clone(),
        artifact_receipt_sha256: value.artifact_receipt_sha256.clone(),
    }
}

fn event_ref(value: &InteractionEvent) -> InteractionEventRef {
    InteractionEventRef {
        event_id: value.event_id.clone(),
        event_sha256: value.event_sha256.clone(),
    }
}

fn execution_ref(value: &ExecutionReceipt) -> ExecutionReceiptRef {
    ExecutionReceiptRef {
        execution_receipt_id: value.execution_receipt_id.clone(),
        execution_receipt_sha256: value.execution_receipt_sha256.clone(),
    }
}

fn validate_common_context(
    closure: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    let base = &closure.capability_invocations[0];
    let artifacts_match = closure
        .artifact_receipts
        .iter()
        .all(|value| value.bindings == base.bindings && value.task_binding == base.task_binding);
    let invocations_match = closure
        .capability_invocations
        .iter()
        .all(|value| value.bindings == base.bindings && value.task_binding == base.task_binding);
    let events_match = closure
        .interaction_events
        .iter()
        .all(|value| value.bindings == base.bindings && value.task_binding == base.task_binding);
    let receipts_match = closure
        .execution_receipts
        .iter()
        .all(|value| value.bindings == base.bindings && value.task_binding == base.task_binding);
    if artifacts_match && invocations_match && events_match && receipts_match {
        Ok(())
    } else {
        Err(invalid(
            "every record must carry identical bindings and TaskBinding",
        ))
    }
}

fn retry_static_matches(first: &CapabilityInvocation, next: &CapabilityInvocation) -> bool {
    first.bindings == next.bindings
        && first.capability == next.capability
        && first.capability_grant_ref == next.capability_grant_ref
        && first.correlation_id == next.correlation_id
        && first.declared_output_slots == next.declared_output_slots
        && first.idempotency_key == next.idempotency_key
        && first.input_artifact_receipt_refs == next.input_artifact_receipt_refs
        && first.requested_action_sha256 == next.requested_action_sha256
        && first.subject == next.subject
        && first.task_binding == next.task_binding
}

fn validate_retry_chain(
    closure: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    if closure.capability_invocations.len() != closure.execution_receipts.len() {
        return Err(invalid(
            "each invocation requires exactly one ExecutionReceipt",
        ));
    }
    let base = &closure.capability_invocations[0];
    for (index, (invocation, receipt)) in closure
        .capability_invocations
        .iter()
        .zip(&closure.execution_receipts)
        .enumerate()
    {
        let attempt = i64::try_from(index + 1).map_err(|_| invalid("attempt overflow"))?;
        if invocation.attempt != attempt || receipt.attempt != attempt {
            return Err(invalid(
                "invocations and receipts must be contiguous attempts 1..N",
            ));
        }
        if !retry_static_matches(base, invocation) {
            return Err(invalid(
                "retry invocations must preserve the complete static request",
            ));
        }
        let prior = index
            .checked_sub(1)
            .map(|prior| execution_ref(&closure.execution_receipts[prior]));
        if invocation.prior_execution_receipt_ref != prior
            || receipt.prior_execution_receipt_ref != prior
        {
            return Err(invalid("prior receipt must be the preceding attempt"));
        }
        validate_retry_time(closure, index, invocation)?;
    }
    Ok(())
}

fn validate_retry_time(
    closure: &KernelOperationalReferenceClosure,
    index: usize,
    invocation: &CapabilityInvocation,
) -> Result<(), KernelOperationalContractError> {
    if index == 0 {
        return Ok(());
    }
    let prior = &closure.execution_receipts[index - 1];
    if prior.outcome == "succeeded" {
        return Err(invalid("a succeeded execution cannot be retried"));
    }
    if invocation.requested_at_unix_ms < prior.ended_at_unix_ms {
        Err(invalid("retry request precedes the prior attempt end"))
    } else {
        Ok(())
    }
}

fn receipt_map(
    values: &[ArtifactReceipt],
) -> Result<HashMap<&str, &ArtifactReceipt>, KernelOperationalContractError> {
    let mut result = HashMap::with_capacity(values.len());
    for value in values {
        if result
            .insert(value.artifact_receipt_id.as_str(), value)
            .is_some()
        {
            return Err(invalid("duplicate ArtifactReceipt identity"));
        }
    }
    Ok(result)
}

fn sorted_artifacts(
    values: impl IntoIterator<Item = ArtifactRef>,
) -> Result<Vec<ArtifactRef>, KernelOperationalContractError> {
    let mut keyed = Vec::new();
    for value in values {
        keyed.push((wire::canonical_typed(&value)?, value));
    }
    keyed.sort_by(|left, right| left.0.cmp(&right.0));
    Ok(keyed.into_iter().map(|(_, value)| value).collect())
}

fn resolve_inputs(
    invocation: &CapabilityInvocation,
    receipt: &ExecutionReceipt,
    available: &HashMap<&str, &ArtifactReceipt>,
    used: &mut BTreeSet<String>,
) -> Result<(), KernelOperationalContractError> {
    let mut artifacts = Vec::with_capacity(invocation.input_artifact_receipt_refs.len());
    let mut seen = BTreeSet::new();
    for reference in &invocation.input_artifact_receipt_refs {
        let member = available
            .get(reference.artifact_receipt_id.as_str())
            .ok_or_else(|| invalid("input ArtifactReceipt reference is unresolved"))?;
        if artifact_receipt_ref(member) != *reference {
            return Err(invalid(
                "input ArtifactReceipt reference digest is unresolved",
            ));
        }
        if member.receipt_role != "declared_input"
            || member.created_at_unix_ms > invocation.requested_at_unix_ms
        {
            return Err(invalid("invocation input role or time is invalid"));
        }
        if !seen.insert(member.artifact.clone()) {
            return Err(invalid(
                "invocation inputs must project distinct ArtifactRefs",
            ));
        }
        used.insert(member.artifact_receipt_id.clone());
        artifacts.push(member.artifact.clone());
    }
    if receipt.input_artifacts == sorted_artifacts(artifacts)? {
        Ok(())
    } else {
        Err(invalid(
            "ExecutionReceipt input_artifacts must exactly project inputs",
        ))
    }
}

fn validate_outputs(
    invocation: &CapabilityInvocation,
    receipt: &ExecutionReceipt,
    values: &[ArtifactReceipt],
    used: &mut BTreeSet<String>,
) -> Result<(), KernelOperationalContractError> {
    let reference = invocation_ref(invocation);
    let outputs: Vec<_> = values
        .iter()
        .filter(|value| value.producer_invocation_ref.as_ref() == Some(&reference))
        .collect();
    let mut slots = Vec::with_capacity(outputs.len());
    let mut refs = Vec::with_capacity(outputs.len());
    for output in outputs {
        if output.created_at_unix_ms < receipt.started_at_unix_ms
            || output.created_at_unix_ms > receipt.ended_at_unix_ms
        {
            return Err(invalid(
                "output ArtifactReceipt time is outside its execution",
            ));
        }
        used.insert(output.artifact_receipt_id.clone());
        slots.push(output.slot.clone());
        refs.push(artifact_receipt_ref(output));
    }
    slots.sort();
    refs.sort_by(|left, right| left.artifact_receipt_id.cmp(&right.artifact_receipt_id));
    if slots.windows(2).any(|pair| pair[0] == pair[1]) || slots != invocation.declared_output_slots
    {
        return Err(invalid(
            "declared_output_slots must exactly cover output receipts",
        ));
    }
    if refs == receipt.output_artifact_receipt_refs {
        Ok(())
    } else {
        Err(invalid(
            "ExecutionReceipt must reference every exact output receipt",
        ))
    }
}

fn events_for<'a>(
    invocation: &CapabilityInvocation,
    receipt: &ExecutionReceipt,
    values: &'a [InteractionEvent],
) -> Result<Vec<&'a InteractionEvent>, KernelOperationalContractError> {
    let reference = invocation_ref(invocation);
    let members: Vec<_> = values
        .iter()
        .filter(|value| value.invocation_ref == reference)
        .collect();
    let mut refs: Vec<InteractionEventRef> = Vec::with_capacity(members.len());
    for (index, event) in members.iter().enumerate() {
        if event.logical_sequence != i64::try_from(index + 1).unwrap_or(i64::MAX) {
            return Err(invalid(
                "events must be contiguous sequences per invocation",
            ));
        }
        let cause = index.checked_sub(1).map(|prior| refs[prior].clone());
        if event.causation_event_ref != cause {
            return Err(invalid("event causation must be immediately preceding"));
        }
        if event.occurred_at_unix_ms < receipt.started_at_unix_ms
            || event.occurred_at_unix_ms > receipt.ended_at_unix_ms
        {
            return Err(invalid("event time is outside its execution"));
        }
        if index > 0 && event.occurred_at_unix_ms < members[index - 1].occurred_at_unix_ms {
            return Err(invalid(
                "event times must be nondecreasing by logical sequence",
            ));
        }
        refs.push(event_ref(event));
    }
    if refs == receipt.event_refs {
        Ok(members)
    } else {
        Err(invalid(
            "ExecutionReceipt event_refs must exactly cover its events",
        ))
    }
}

fn validate_attempt<'a>(
    invocation: &CapabilityInvocation,
    receipt: &ExecutionReceipt,
    artifact_map: &HashMap<&str, &ArtifactReceipt>,
    artifacts: &[ArtifactReceipt],
    events: &'a [InteractionEvent],
    used: &mut BTreeSet<String>,
) -> Result<Vec<&'a InteractionEvent>, KernelOperationalContractError> {
    if receipt.invocation_ref != invocation_ref(invocation) {
        return Err(invalid(
            "ExecutionReceipt must reference its matching invocation",
        ));
    }
    if receipt.correlation_id != invocation.correlation_id {
        return Err(invalid("receipt and invocation correlation_id differ"));
    }
    if receipt.started_at_unix_ms < invocation.requested_at_unix_ms {
        return Err(invalid("execution starts before its invocation request"));
    }
    resolve_inputs(invocation, receipt, artifact_map, used)?;
    validate_outputs(invocation, receipt, artifacts, used)?;
    let members = events_for(invocation, receipt, events)?;
    if members
        .iter()
        .any(|event| event.correlation_id != invocation.correlation_id)
    {
        return Err(invalid("event and invocation correlation_id differ"));
    }
    Ok(members)
}

fn validate_artifact_inventory(
    closure: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    let mut created: HashMap<ArtifactRef, Vec<i64>> = HashMap::new();
    let mut projected = BTreeSet::new();
    for receipt in &closure.artifact_receipts {
        created
            .entry(receipt.artifact.clone())
            .or_default()
            .push(receipt.created_at_unix_ms);
        projected.insert(receipt.artifact.clone());
    }
    for event in &closure.interaction_events {
        for artifact in &event.artifact_refs {
            let nonfuture = created.get(artifact).is_some_and(|times| {
                times
                    .iter()
                    .any(|created| *created <= event.occurred_at_unix_ms)
            });
            if !nonfuture {
                return Err(invalid(
                    "Event ArtifactRef needs a non-future included receipt",
                ));
            }
            projected.insert(artifact.clone());
        }
    }
    for receipt in &closure.execution_receipts {
        projected.extend(receipt.input_artifacts.iter().cloned());
    }
    if closure.artifacts == sorted_artifacts(projected)? {
        Ok(())
    } else {
        Err(invalid(
            "closure artifacts must exactly equal all referenced ArtifactRefs",
        ))
    }
}

pub(super) fn validate_reference_graph(
    closure: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    validate_common_context(closure)?;
    validate_retry_chain(closure)?;
    let artifact_map = receipt_map(&closure.artifact_receipts)?;
    let mut used = BTreeSet::new();
    let mut flattened = Vec::with_capacity(closure.interaction_events.len());
    for (invocation, receipt) in closure
        .capability_invocations
        .iter()
        .zip(&closure.execution_receipts)
    {
        flattened.extend(validate_attempt(
            invocation,
            receipt,
            &artifact_map,
            &closure.artifact_receipts,
            &closure.interaction_events,
            &mut used,
        )?);
    }
    if flattened != closure.interaction_events.iter().collect::<Vec<_>>() {
        return Err(invalid(
            "interaction_events must be ordered by attempt then sequence",
        ));
    }
    if used.len() != artifact_map.len() {
        return Err(invalid("ArtifactReceipt inventory contains an orphan"));
    }
    validate_artifact_inventory(closure)
}
